### PR TITLE
Add datetime functions `FROM_UNIXTIME` and `UNIX_TIMESTAMP`

### PR DESCRIPTION
**Pull request by Yury-Fridlyand**
## [FROM_UNIXTIME](https://dev.mysql.com/doc/refman/8.0/en/date-and-time-functions.html#function_from-unixtime)
* `(DOUBLE) -> DATETIME`
* `(DOUBLE, STRING) -> STRING`

### Implementation details
* For string formatting `DATE_FORMAT` function is used
* Function returns 0 for values outside of range `1970-01-01 00:00:00` - `3001-01-18 23:59:59.999999`

### NOTES
* Returning an integer in numeric context is not supported.
* Operations with `DOUBLE` affected by [Floating Point Imprecision](https://timothybramlett.com/floating-point-imprecision.html)
<details>
<summary>Example</summary>

```
opensearchsql> select FROM_UNIXTIME(1220249547.12);
fetched rows / total rows = 1/1
+--------------------------------+
| FROM_UNIXTIME(1220249547.12)   |
|--------------------------------|
| 2008-09-01 06:12:27.119999885  |
+--------------------------------+
```
</details>

### Test samples
<details>
<summary>SQL</summary>

```sql
select FROM_UNIXTIME(1337.42), FROM_UNIXTIME(123235435, '%T %M %D, %Y');
```
```
+--------------------------+--------------------------------------------+
| FROM_UNIXTIME(1337.42)   | FROM_UNIXTIME(123235435, '%T %M %D, %Y')   |
|--------------------------+--------------------------------------------|
| 1970-01-01 00:22:17.42   | 08:03:55 November 27th, 1973               |
+--------------------------+--------------------------------------------+
```
</details>

<details>
<summary>PPL</summary>

```sql
source=date | top 1 date_keyword | eval `FROM_UNIXTIME(1337.42)` = FROM_UNIXTIME(1337.42), `FROM_UNIXTIME(123235435, '%T %M %D, %Y')` = FROM_UNIXTIME(123235435, '%T %M %D, %Y') | fields `FROM_UNIXTIME(1337.42)`, `FROM_UNIXTIME(123235435, '%T %M %D, %Y')`;
```
```
+--------------------------+--------------------------------------------+
| FROM_UNIXTIME(1337.42)   | FROM_UNIXTIME(123235435, '%T %M %D, %Y')   |
|--------------------------+--------------------------------------------|
| 1970-01-01 00:22:17.42   | 08:03:55 November 27th, 1973               |
+--------------------------+--------------------------------------------+
```
</details>

## [UNIX_TIMESTAMP](https://dev.mysql.com/doc/refman/8.0/en/date-and-time-functions.html#function_unix-timestamp)
* `() -> LONG`
* `DATE -> DOUBLE`
* `DATETIME -> DOUBLE`
* `TIMESTAMP -> DOUBLE`
* `DOUBLE -> DOUBLE`

### Implementation details
* First function signature calls for `LocalDateTime.now()` without caching the value. Sequential calls in one query might return different values.
* Function returns 0 for values outside of range `1970-01-01 00:00:00` - `3001-01-18 23:59:59.999999`.

### NOTES
* Operations with `DOUBLE` affected by [Floating Point Imprecision](https://timothybramlett.com/floating-point-imprecision.html)
<details>
<summary>Example</summary>

```
opensearchsql> select UNIX_TIMESTAMP(20771122143845.2345);
fetched rows / total rows = 1/1
+---------------------------------------+
| UNIX_TIMESTAMP(20771122143845.2345)   |
|---------------------------------------|
| 3404817524.234375                     |
+---------------------------------------+
```
</details>

### Test samples
<details>
<summary>SQL</summary>

```sql
select UNIX_TIMESTAMP(), UNIX_TIMESTAMP(20771122143845), UNIX_TIMESTAMP(881226), UNIX_TIMESTAMP(881226231058.123), UNIX_TIMESTAMP(DATE('1995-05-06')), UNIX_TIMESTAMP(TIMESTAMP('2022-08-09 12:16:12')), UNIX_TIMESTAMP(ADDDATE(DATE('2020-08-26'), INTERVAL 1 HOUR));
```
```
UNIX_TIMESTAMP()                                             | 1662664900
UNIX_TIMESTAMP(20771122143845)                               | 3404817525.0
UNIX_TIMESTAMP(881226)                                       | 599097600.0
UNIX_TIMESTAMP(881226231058.123)                             | 599181057.1230469
UNIX_TIMESTAMP(DATE('1995-05-06'))                           | 799718400.0
UNIX_TIMESTAMP(TIMESTAMP('2022-08-09 12:16:12'))             | 1660047372.0
UNIX_TIMESTAMP(ADDDATE(DATE('2020-08-26'), INTERVAL 1 HOUR)) | 1598403600.0
```
</details>

<details>
<summary>PPL</summary>

```sql
source=date | eval `UNIX_TIMESTAMP()` = UNIX_TIMESTAMP(), `UNIX_TIMESTAMP(20771122143845)` = UNIX_TIMESTAMP(20771122143845), `UNIX_TIMESTAMP(881226)` = UNIX_TIMESTAMP(881226), `UNIX_TIMESTAMP(881226231058.123)` = UNIX_TIMESTAMP(881226231058.123), `UNIX_TIMESTAMP(DATE('1995-05-06'))` = UNIX_TIMESTAMP(DATE('1995-05-06')), `UNIX_TIMESTAMP(TIMESTAMP('2022-08-09 12:16:12'))` = UNIX_TIMESTAMP(TIMESTAMP('2022-08-09 12:16:12')), `UNIX_TIMESTAMP(ADDDATE(DATE('2020-08-26'), INTERVAL 1 HOUR))` = UNIX_TIMESTAMP(ADDDATE(DATE('2020-08-26'), INTERVAL 1 HOUR)) | fields `UNIX_TIMESTAMP()`, `UNIX_TIMESTAMP(20771122143845)`, `UNIX_TIMESTAMP(881226)`, `UNIX_TIMESTAMP(881226231058.123)`, `UNIX_TIMESTAMP(DATE('1995-05-06'))`, `UNIX_TIMESTAMP(TIMESTAMP('2022-08-09 12:16:12'))`, `UNIX_TIMESTAMP(ADDDATE(DATE('2020-08-26'), INTERVAL 1 HOUR))`;
```
```
UNIX_TIMESTAMP()                                             | 1662664900
UNIX_TIMESTAMP(20771122143845)                               | 3404817525.0
UNIX_TIMESTAMP(881226)                                       | 599097600.0
UNIX_TIMESTAMP(881226231058.123)                             | 599181057.1230469
UNIX_TIMESTAMP(DATE('1995-05-06'))                           | 799718400.0
UNIX_TIMESTAMP(TIMESTAMP('2022-08-09 12:16:12'))             | 1660047372.0
UNIX_TIMESTAMP(ADDDATE(DATE('2020-08-26'), INTERVAL 1 HOUR)) | 1598403600.0
```
</details>
 
## Check List
- [x] New functionality includes testing.
  - [x] All tests pass, including unit test, integration test and doctest
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
  - [x] New functionality has user manual doc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java [hunk 1]
```diff
@@ -721,6 +784,103 @@ private ExprValue exprWeek(ExprValue date, ExprValue mode) {
         CalendarLookup.getWeekNumber(mode.integerValue(), date.dateValue()));
   }
 
+  private ExprValue unixTimeStamp() {
+    return new ExprLongValue(Instant.now().getEpochSecond());
```
Comment (dai-chen) [inline core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java:788]: Should this depend on `now()` implementation in https://github.com/opensearch-project/sql/pull/754?

File: core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java [hunk 2]
```diff
@@ -519,6 +553,35 @@ private ExprValue exprFromDays(ExprValue exprValue) {
...
   }
 
+  private ExprValue exprFromUnixTime(ExprValue time) {
+    if (0 > time.doubleValue()) {
+      return ExprNullValue.of();
+    }
+    // According to MySQL documentation:
+    //     effective maximum is 32536771199.999999, which returns '3001-01-18 23:59:59.999999' UTC.
+    //     Regardless of platform or version, a greater value for first argument than the effective
+    //     maximum returns 0.
```
Comment (dai-chen) [inline core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java:563]: returns null ?

File: core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java [hunk 3]
```diff
@@ -721,6 +784,103 @@ private ExprValue exprWeek(ExprValue date, ExprValue mode) {
...
+            .appendValueReduced(YEAR, 2, 2, 1970)
+            .appendPattern("MMddHHmmss")
+            .toFormatter()
+            .withResolverStyle(ResolverStyle.STRICT);
+
+        var dateTimeFormatLongYear = new DateTimeFormatterBuilder()
+            .appendValue(YEAR,4)
+            .appendPattern("MMddHHmmss")
+            .toFormatter()
+            .withResolverStyle(ResolverStyle.STRICT);
```
Comment (dai-chen) [inline core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java:843]: Should we make all formatters static or move it to some utility class?

File: core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java (hunk 2 above)
Comment (Yury-Fridlyand) [reply core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java:563]: Documentation says 0, actually returns NULL.
```
mysql> select from_unixtime(32536771199.999999 + 1);
+---------------------------------------+
| from_unixtime(32536771199.999999 + 1) |
+---------------------------------------+
| NULL                                  |
+---------------------------------------+
1 row in set (0.00 sec)
```

File: core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java (hunk 1 above)
Comment (Yury-Fridlyand) [reply core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java:788]: How can I do this properly for `unix_timestamp()` without arguments only?
Or I can cache all `unix_timestamp` calls.

Comment (dai-chen) [reply core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java:788]: Sorry, I just finished that PR and may misunderstood. In that PR, we treated `now()` as constant function which is replaced by a "cached" constant value. So no matter how many call to now(), they are replaced by same literal value during analysis. Ignore my previous comment if this is not the case for unix_timestamp().

File: core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java (hunk 2 above)
Comment (dai-chen) [reply core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java:563]: I see. The comment is copied from MySQL doc.

File: core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java (hunk 3 above)
Comment (Yury-Fridlyand) [reply core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java:843]: We have class `OpenSearchDateFormatters`, but it is located in `opensearch` and not accessible in `core`. Do you want me to move it or use `DateTimeUtils` instead?

Comment (dai-chen) [reply core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java:843]: Probably we can move it to `core` and use it to hold all common date formatters? Then `OpenSearchExprValueFactory` and others can choose which formatter is needed by themselves?

**Review by dai-chen — APPROVED** [review]
LGTM. Thanks for the changes!

**Comment by Yury-Fridlyand** [comment]
Maybe I need to rename `DateFormatters.java` to `DateTimeFormatters.java`?

**Comment by dai-chen** [comment]
> Maybe I need to rename `DateFormatters.java` to `DateTimeFormatters.java`?

Sure, we can do it now or later. Thanks!

**Review by dai-chen — DISMISSED** [review]

**Comment by Yury-Fridlyand** [comment]
Please, ~re-approve~ re-review. I rebased the branch to fix merge conflicts.

**Review by MaxKsyunz — APPROVED** [review]

**Review by penghuo — APPROVED** [review]

**Review by dai-chen — APPROVED** [review]

**Pull request merged**